src/tests/core/txs/test_wallet.c

```diff
@@ -0,0 +1,284 @@
+      options->in_total = 17
+  );
+  mu_assert(
+      "Options has wrong out_total",
+      options->out_total = 15
+  );
+
+
+  mu_assert(
+      "New tx has wrong number of inputs",
+      tx->num_inputs == 2
+  );
+  i = 0;
+  for (map_value = wallet_pool; map_value != NULL; map_value = map_value->hh.next) {
+    mu_assert(
+        "Input has wrong vout",
+        map_value->id.vout == tx->inputs[i].prev_utxo_output
+    );
+    mu_assert(
+        "Input has wrong tx hash",
+        memcmp(map_value->id.tx_hash, tx->inputs[i].prev_tx_id, TX_HASH_LEN) == 0
+    );
+    mu_assert(
+        "Input has wrong tx pub_key",
+        mbedtls_ecp_point_cmp(
+          &(map_value->entry->key_pair->private_Q),
+          tx->inputs[i].pub_key
+        ) == 0
+    );
+    mu_assert(
+        "Input has wrong tx signature",
+        memcmp(tx->inputs[i].signature, empty_sig, SIGNATURE_LEN) == 0
+    );
+
+    mu_assert(
+        "Wrong input key pair returned",
+        map_value->entry->key_pair == ret_keys[i]
+    );
+    mu_assert(
+        "Hash map entry spent not set",
+        map_value->entry->spent == 1
+    );
+
+    i++;
+  }
+
+  _free_options(options);
+  _free_wallet_pool();
+  _free_tx(tx);
+  free(ret_keys);
+
+  return NULL;
+}
+
+static char *test_build_outputs() {
+  TxOptions *options;
+  Transaction *tx;
+  mbedtls_ecdsa_context *key_pair;
+  unsigned char key_hash[PUB_KEY_HASH_LEN];
+
+  options = _make_options();
+  options->in_total = 17;
+  options->out_total = 15;
+  tx = malloc(sizeof(Transaction));
+
+  key_pair = build_outputs(tx, options);
+  hash_pub_key(key_hash, key_pair);
+
+  for (size_t i = 0; i < options->num_dests; i++) {
+    mu_assert(
+        "Output has wrong amt",
+        options->dests[i].amt == tx->outputs[i].amt
+    );
+    mu_assert(
+        "Output has wrong pub key hash",
+        memcmp(
+          options->dests[i].public_key_hash,
+          tx->outputs[i].public_key_hash,
+          PUB_KEY_HASH_LEN
+        ) == 0
+    );
+  }
+
+  mu_assert(
+      "Self output has wrong amt",
+      tx->outputs[options->num_dests].amt == 2
+  );
+  mu_assert(
+      "Self output has wrong pub key hash",
+      memcmp(
+        tx->outputs[options->num_dests].public_key_hash,
+        key_hash,
+        PUB_KEY_HASH_LEN
+      ) == 0
+  );
+
+  _free_options(options);
+  _free_tx(tx);
+  mbedtls_ecp_keypair_free(key_pair);
+
+  return NULL;
+}
+
+static char *test_sign_tx() {
+  TxOptions *options;
+  Transaction *tx;
+  mbedtls_ecdsa_context **ret_keys;
+  unsigned char tx_hash[TX_HASH_LEN];
+
+  wallet_init();
+
+  _populate_wallet_pool();
+  options = _make_options();
+  tx = malloc(sizeof(Transaction));
+  tx->num_outputs = 0;
+  tx->outputs = NULL;
+  ret_keys = build_inputs(tx, options);
+  hash_tx(tx_hash, tx);
+
+  sign_tx(tx, ret_keys);
+  for (size_t i = 0; i < tx->num_inputs; i++) {
+    mu_assert(
+        "Input signature invalid",
+        mbedtls_ecdsa_read_signature(
```

Comment:
Wrapping the read signature as a validate function might save us time later

Reply:
For sure. Was actually planning to just copy this when we actually had to write it haha, probably worth doing now